src/Notes.h

```diff
@@ -262,5 +267,18 @@ class NotePairs: public RefCountable
     Entries entries; ///< The key/value pair entries
 };
 
+
+template <class Cond>
+void
+NotePairs::replaceOrAddIf(const NotePairs *src, const Cond cond)
+{
+    for (const auto &e: src->entries) {
+        if (cond(e)) {
+            remove(e->name());
+            add(e->name(), e->value());
```

Comment:
> Does that helper violate any documentation or written advice?

I could not find such limitation. Auth helper in Auth::User::absorb() avoids copying kv-pair duplicates (i.e., matching both key and value), but it does not say that the helper should not send such duplicates.

> Does the official helper response parsing code in Squid remove the same-name annotations returned by such a helper?

The helper parsing code is confined in Helper::Reply::parseResponseKeys() and it does not filter out any duplicates.

> If replaceOrAdd() adds same-name pairs to the NotePairs, what happens to them in the code that uses/reads those NotePairs?

`%note` logs all same-name pairs from helper. `note` ACL (ACLNoteStrategy::matchNotes()) searches through all notes for the first occurrence of the specified key (nil value) or for the first occurrence of the specified key+value.

> this code effectively adds only the last such entry, while replaceOrAdd() adds all such entries.

replaceOrAddIf() substituted the old logic in UpdateRequestNotes(), where 1) all old clt_conn_tag duplicates are removed and 2) one new clt_conn_tag=value is added (the one returned by NotePairs::findFirst()). However, I see now that the new code is different - it adds the last occurrence, while the old code added the first one. On the other hand, we can say that this logic did not allow clt_conn_tag (in the client annotation) to have more that one value in the first place. Documentation confirms this, saying that "A helper may update the TAG during subsequent requests by returning a new kv-pair", meaning that the value will be overwritten, IMO. So, if we are going to continue having two kinds of annotations, i.e., 1) client connection (with unique keys) and 2) transaction (with duplicate keys), we would need two different methods, i.e., replaceOrAddIf() (or equivalent) and replaceOrAdd().

Reply:
Thank you for this research. I acknowledge that we actually have _three_ functions to worry about: old replaceOrAdd(), new replaceOrAddIf(), and old UpdateRequestNotes(). They all have slightly different semantics!

> Documentation confirms this, saying that "A helper may update the TAG during subsequent requests by returning a new kv-pair", meaning that the value will be overwritten

That documentation is about two transactions, each sending one annotation. The situation is clear(er) there -- the new same-name annotation will "naturally" overwrite the old one.

Our current use case is a single transaction with two same-name annotations. I wish the official code would not add or would not use all of them, but it does! Should we claim that if a helper sends two same-name annotations, then Squid should warn and use just the last one because it is more likely to be a helper bug rather than a true multi-valued annotation? Helpers that want to return a multi-valued annotation should use a single annotation that has commas or other delimiters in its value. ACLNoteStrategy::matchNotes() already supports such multi-valued annotations AFAICT.

Other than Squid code simplification and Squid speed, can you think of any reason to ban helpers from sending multiple same-name annotations? I cannot think of any place where official Squid clearly mishandles such annotations except that UpdateRequestNotes() silently ignores all but the first one. Do you think that single problem and simplicity/speed are enough to ban something (rare) that (otherwise) seems to work OK?

Reply:
> can you think of any reason to ban helpers from sending multiple same-name annotations?

I would also put a little different question: is there any helper-result-processing code dealing with multiple same-name annotations? I could find only a couple of such places (in auth code), where all `message` values are extracted by NotePairs::find(). Other (known) keys are extracted by means of NotePairs::findFirst().
So, if (at least) we are going to keep supporting this multiple 'message' values, we would not be able to completely ban same-name annotations. On the other hand, if this is not important, I do not see any other (good) reason why not to ban. I agree that the comma-separated mutli-value syntax (as in note ACL) would be a better alternative for this purpose.

Reply:
>  ban helpers from sending multiple same-name annotations

Apart from helpers, there are other sources of same-name annotations. I tried to list them all below:

* 'note' directive (Config.notes)
* 'adaptation_meta' (Adaptation::History::metaHeaders)
* annotate_transaction ACL (in its 'addition' format)
* annotate_client ACL (in its 'addition' format)
* helpers, however Squid extracts only the first key-value pair for all reserved fields, except the 'message' field (while transaction annotation still receives all same-key values).

Also, although 'note' ACL documentation mentions eCAP responses as another source, I could not find the corresponding code. Its ACLNoteStrategy::match() analyzes two sources: request->notes() and ah->metaHeaders (coming from adaptation_meta). Response headers from helpers are stored in History::allMeta, but they are not   copied into any of these two Notes.

Reply:
> * annotate_transaction ACL (in its 'addition' format)
> * annotate_client ACL (in its 'addition' format)

Do you mean that these ACLs can supply same-name annotations only when using addition format? Or that they can append to existing same-name annotations using the addition format? Or something else?

Does evaluating the following ACL adds two values (to an annotation named "noteA")?

```
# adds two values?
acl addTwoValues annotate_transaction noteA+=1 noteA+=2
```

```
# adds just one value? 2?
acl addOneValue annotate_transaction noteA=1 noteA=2
```

Any warnings/errors when either of the above ACLs is used?

Reply:
> Also, although 'note' ACL documentation mentions eCAP responses as another source, I could not find the corresponding code.

I think this is a bug we fixed in commit 88ca55c. We need to make that fix official, but doing that is outside this PR scope.

Reply:
> > Also, although 'note' ACL documentation mentions eCAP responses as another source, I could not find the corresponding code.
> 
> I think this is a bug we fixed in commit [88ca55c](https://github.com/measurement-factory/squid/commit/88ca55ca841f9c72f72e594e3d049891b73f43bf). We need to make that fix official, but doing that is outside this PR scope.

AFAICT, that fix adds the last of all same-name annotations. See Adaptation::History::recordReceivedAnnotations().

Reply:
> Do you mean that these ACLs can supply same-name annotations only when using addition format? 

These ACLs add new same-name annotations (or append to existing same-name annotations) only in the addition (mhAppend) format (Note::updateNotePairs()). In the replace format (mhReplace) only the last value specified in an ACL configuration will go into the annotation (i.e., all previously configured same-name values in the ACL will be ignored and already existing same-name values in the annotation will be removed from the annotation (Notes::parseKvPair()).

> \# adds two values?
> acl addTwoValues annotate_transaction noteA+=1 noteA+=2

Yes.

> \# adds just one value? 2?
> acl addOneValue annotate_transaction noteA=1 noteA=2

Yes, and removes already-existing values with the noteA key.

> Any warnings/errors when either of the above ACLs is used?

```
WARNING: annotation configuration with key [...] already exists and will be overwritten
```

Reply:
> that fix adds the last of all same-name annotations. See Adaptation::History::recordReceivedAnnotations().

Yes, recordReceivedAnnotations() and replaceOrAddIf() work alike in this aspect.

Reply:
```
WARNING: annotation configuration with key [...] already exists and will be overwritten
```

Just FTR: The above configuration-time WARNING is for mhReplace mode only (because mhAppend cases do not overwrite "existing" same-name annotations).